jgrapht-core/src/main/java/org/jgrapht/alg/shortestpath/FloydWarshallShortestPaths.java

```diff
@@ -0,0 +1,437 @@
+     * 
+     * @return the graph on which this algorithm operates
+     */
+    public Graph<V, E> getGraph()
+    {
+        return graph;
+    }
+
+    /**
+     * Get the total number of shortest paths.
+     * 
+     * @return total number of shortest paths
+     */
+    public int getShortestPathsCount()
+    {
+        lazyCalculateMatrix();
+
+        return nShortestPaths;
+    }
+
+    /**
+     * Get the length of a shortest path.
+     *
+     * @param a first vertex
+     * @param b second vertex
+     *
+     * @return shortest distance between a and b
+     */
+    public double shortestDistance(V a, V b)
+    {
+        if (!graph.containsVertex(a)) {
+            throw new IllegalArgumentException("graph must contain the source vertex");
+        }
+        if (!graph.containsVertex(b)) {
+            throw new IllegalArgumentException("graph must contain the sink vertex");
+        }
+
+        lazyCalculateMatrix();
+
+        return d[vertexIndices.get(a)][vertexIndices.get(b)];
+    }
+
+    /**
+     * Compute the diameter of the graph.
+     * 
+     * @return the diameter (longest of all the shortest paths) computed for the graph. If the graph
+     *         contains no vertices, return {@link Double#NaN}. If there is no path between any two
+     *         vertices, return {@link Double#POSITIVE_INFINITY}.
+     */
+    public double getDiameter()
+    {
+        lazyCalculateMatrix();
+
+        if (!Double.isNaN(diameter)) {
+            return diameter;
+        }
+
+        int n = vertices.size();
+        if (n > 0) {
+            diameter = 0.0;
+            for (int i = 0; i < n; i++) {
+                for (int j = 0; j < n; j++) {
+                    diameter = Double.max(diameter, d[i][j]);
+                }
+            }
+        }
+        return diameter;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public GraphPath<V, E> getPath(V a, V b)
+    {
+        if (!graph.containsVertex(a)) {
+            throw new IllegalArgumentException("graph must contain the source vertex");
+        }
+        if (!graph.containsVertex(b)) {
+            throw new IllegalArgumentException("graph must contain the sink vertex");
+        }
+
+        lazyCalculateMatrix();
+
+        int v_a = vertexIndices.get(a);
+        int v_b = vertexIndices.get(b);
+
+        if (backtrace[v_a][v_b] == null) { // No path exists
+            return createEmptyPath(a, b);
+        }
+
+        // Reconstruct the path
+        List<E> edges = new ArrayList<>();
+        V u = a;
+        while (!u.equals(b)) {
+            int v_u = vertexIndices.get(u);
+            E e = TypeUtil.uncheckedCast(backtrace[v_u][v_b], null);
+            edges.add(e);
+            u = Graphs.getOppositeVertex(graph, e, u);
+        }
+        return new GraphWalk<>(graph, a, b, null, edges, d[v_a][v_b]);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public SingleSourcePaths<V, E> getPaths(V source)
```

Comment:
This doesn't seem ideal. This method transforms the internal datastructure of FloydWarshall (its distance matrix) to the datastructure of TreeSingleSourcePaths (a map with predecessors) which is pure overhead. These data structures encode the same, but in a different format. Whenever the user queries a path from TreeSingleSourcePaths, the getPath(V sink) method in TreeSingleSourcePaths performs the same steps as the getPath(V source, V sink) in FloydWarshall would do. Consequently, if the user wants to query the paths from a single source u to various sink vertices v, the user would be better off to repeatedly invoke the getPath(V u, V v) method than the getPath(V v) method.

For this particular class, it would be better if it would directly implement SingleSourcePaths. An invocation to getPath(V v) can than simply return getPath(V u, V v)